R/find_and_remove_duplicates.R

```diff
@@ -0,0 +1,143 @@
+                              remove         = NULL) {
+
+  # setting up the variables below to NULL to avoid linters
+  row_id <- NULL # nolint: object_usage_linter
+  report <- attr(data, "report")
+  dat    <- data
+
+  # get the target column names
+  target_columns <- get_target_column_names(dat, target_columns,
+                                            cols = NULL)
+
+  # find duplicates
+  dups       <- find_duplicates(dat, target_columns)
+  tmp_report <- attr(dups, "report")
+  if ("duplicated_rows" %in% names(tmp_report) &&
+      nrow(tmp_report[["duplicated_rows"]]) > 0L) {
+    dups     <- tmp_report[["duplicated_rows"]]
+    report   <- c(report, tmp_report)
+    dat      <- dat %>%
+      dplyr::mutate(row_id = seq_len(nrow(dat)))
+  }
+
+  # remove duplicates
+  if (is.null(remove)) {
+    # remove duplicates and keep the first instance of the duplicate in each
+    # duplicate group
+    dat <- dat %>%
+      dplyr::distinct_at({{ target_columns }}, .keep_all = TRUE)
+  } else {
+    # remove duplicates from user specified rows
+    dat <- dat[-remove, ]
+  }
+
+
+  if ("duplicated_rows" %in% names(tmp_report) &&
+      nrow(tmp_report[["duplicated_rows"]]) > 0L) {
+    tmp_target_columns <- c("row_id", target_columns)
+    to_be_removed      <- suppressMessages(dplyr::anti_join(dups, dat) %>%
+        dplyr::select({{ tmp_target_columns }}))
+    report[["removed_duplicates"]] <- to_be_removed
+  }
+
+  attr(dat, which = "report") <- report
+  return(dat)
+}
+
+
+
+#' Identify and return duplicated rows in a data frame or linelist.
+#'
+#' @param data A data frame or linelist.
+#' @param target_columns A vector of columns names or indices to consider when
+#'    looking for duplicates. When the input data is a `linelist` object, this
+#'    parameter can be set to `tags`from which duplicates to be removed.
+#'    Its default value is `NULL`, which considers duplicates across all
+#'    columns.
+#'
+#' @return A data frame or linelist of all duplicated rows with following 2
+#'    additional columns:
+#'    \enumerate{
+#'      \item `row_id`: the indices of the duplicated rows from the input data.
+#'          Users can choose from these indices, which row they consider as
+#'          redundant in each group of duplicates.
+#'      \item `group_id`: a unique identifier associated to each group of
+#'          duplicates.
+#'    }
+#'
+#' @export
+#'
+#' @examples
+#' dups <- find_duplicates(
+#'   data           = readRDS(system.file("extdata", "test_linelist.RDS",
+#'                                        package = "cleanepi")),
+#'   target_columns = c("dt_onset", "dt_report", "sex", "outcome")
+#' )
+#'
+find_duplicates <- function(data, target_columns = NULL) {
+  # get the target column names
+  if (is.null(target_columns)) {
+    target_columns <- names(data)
+  }
+  target_columns   <- get_target_column_names(data, target_columns, cols = NULL)
+
+  # find duplicates
+  num_dups <- row_id <- group_id <- NULL
```

Comment:
A better way to deal with this is to follow the recommendation in https://dplyr.tidyverse.org/articles/in-packages.html#data-masking-and-tidy-selection-notes

Reply:
Accounted for this in commit `6dafba4`.